File: frontend/src/api/time.ts
/**
 * 时间系统API接口
 * GodView v5 时间流逝系统前端API
 */

import { api } from './client'

/**
 * 时间系统配置
 */
export interface TimeSystemConfig {
  world_id: string
  time_scale: number
  time_mode: 'linear' | 'nonlinear' | 'frozen'
  day_length: number
  enable_recording: boolean
  tick_interval_seconds: number
}

/**
 * 时间更新结果
 */
export interface TimeUpdateResult {
  current_time: string
  previous_time: string
  time_delta_minutes: number
  time_scale: number
  tick_count: number
  day_of_week: number
  day_of_week_name: string
  hour: number
  minute: number
  day_phase: string
  season: string
  triggered_events: Array<{
    event_type: string
    data: any
    trigger_time: string
  }>
}

/**
 * 时间系统状态
 */
export interface TimeSystemStatus {
  world_id: string
  time_info: {
    current_time: string
    formatted_time: string
    tick_count: number
    time_scale: number
    day_of_week: number
    day_of_week_name: string
    hour: number
    minute: number
    day_phase: string
    season: string
    month: number
    day: number
    year: number
  }
  time_mode: string
  is_frozen: boolean
}

/**
 * 时间分支信息
 */
export interface TimeBranchInfo {
  id: string
  name: string
  time: string
  tick_count: number
  created_at: string
  scheduled_events_count: number
}

/**
 * 时间点记录
 */
export interface TimePointRecord {
  id: string
  time: string
  tick_count: number
  note: string
  day_phase: string
  season: string
}

/**
 * 时间历史响应
 */
export interface TimeHistoryResponse {
  time_points: TimePointRecord[]
  branches: TimeBranchInfo[]
  current_time: string
  time_mode: string
  tick_count: number
}

/**
 * 创建时间系统
 */
export async function createTimeSystem(worldId: string, config: Partial<TimeSystemConfig> = {}) {
  return api.post('/time/systems/' + worldId, {
    world_id: worldId,
    time_scale: config.time_scale || 1.0,
    time_mode: config.time_mode || 'linear',
    day_length: config.day_length || 24,
    enable_recording: config.enable_recording !== false,
    tick_interval_seconds: config.tick_interval_seconds || 1.0,
  })
}

/**
 * 获取时间系统状态
 */
export async function getTimeSystemStatus(worldId: string): Promise<TimeSystemStatus> {
  const response = await api.get('/time/systems/' + worldId)
  return response.data
}

/**
 * 时间控制操作
 */
export async function controlTime(
  worldId: string,
  action: 'freeze' | 'unfreeze' | 'set_scale' | 'advance' | 'set_mode',
  options?: {
    time_scale?: number
    advance_minutes?: number
    mode?: 'linear' | 'nonlinear' | 'frozen'
  }
) {
  return api.post('/time/control', {
    world_id: worldId,
    action,
    ...options,
  })
}

/**
 * 时间跳跃操作
 */
export async function timeJump(
  worldId: string,
  jumpType: 'forward' | 'backward' | 'to_absolute' | 'to_year',
  options?: {
    target_time?: string
    delta_minutes?: number
    target_year?: number
    note?: string
  }
) {
  return api.post('/time/jump', {
    world_id: worldId,
    jump_type: jumpType,
    ...options,
  })
}

/**
 * 创建时间分支
 */
export async function createTimeBranch(
  worldId: string,
  branchName: string,
  note: string = ''
) {
  return api.post('/time/branches', {
    world_id: worldId,
    branch_name: branchName,
    note,
  })
}

/**
 * 获取时间分支列表
 */
export async function getTimeBranches(worldId: string): Promise<{ success: boolean; world_id: string; branches: TimeBranchInfo[]; count: number }> {
  const response = await api.get('/time/branches/' + worldId)
  return response.data
}

/**
 * 切换时间分支
 */
export async function switchTimeBranch(worldId: string, branchId: string) {
  return api.post('/time/branches/' + worldId + '/switch', null, {
    params: { branch_id: branchId }
  })
}

/**
 * 记录时间点
 */
export async function recordTimePoint(worldId: string, note: string) {
  return api.post('/time/record', {
    world_id: worldId,
    note,
  })
}

/**
 * 获取时间历史
 */
export async function getTimeHistory(worldId: string, limit: number = 100): Promise<TimeHistoryResponse> {
  const response = await api.get('/time/history/' + worldId, {
    params: { limit }
  })
  return response.data
}

// WebSocket 相关类型
export type TimeWebSocketMessage =
  | { type: 'connected'; data: any }
  | { type: 'time_update'; data: TimeUpdateResult }
  | { type: 'time_info'; data: any }
  | { type: 'tick_result'; data: any }
  | { type: 'freeze_toggle'; data: any }
  | { type: 'scale_set'; data: any }

/**
 * WebSocket 时间更新处理器类型
 */
export type TimeUpdateHandler = (update: TimeUpdateResult) => void
export type TimeConnectionHandler = (data: any) => void
export type ErrorHandler = (error: Event) => void

File: app/models/time.py
"""
时间系统相关的数据模型
"""

from datetime import datetime
from typing import Dict, Any, Optional, List
from pydantic import BaseModel, Field
from enum import Enum


class DayPhaseEnum(str, Enum):
    """一天中的时段枚举"""
    MORNING = "早晨"
    FORENOON = "上午"
    NOON = "中午"
    AFTERNOON = "下午"
    EVENING = "傍晚"
    NIGHT = "夜晚"


class SeasonEnum(str, Enum):
    """季节枚举"""
    SPRING = "春"
    SUMMER = "夏"
    AUTUMN = "秋"
    WINTER = "冬"


class TimeModeEnum(str, Enum):
    """时间模式枚举"""
    LINEAR = "linear"
    NONLINEAR = "nonlinear"
    FROZEN = "frozen"


class TimeEventModel(BaseModel):
    """时间事件模型"""
    id: str = Field(..., description="事件ID")
    event_type: str = Field(..., description="事件类型")
    trigger_time: datetime = Field(..., description="触发时间")
    data: Dict[str, Any] = Field(default_factory=dict, description="事件数据")
    recurring: bool = Field(default=False, description="是否循环事件")
    recurrence_interval_minutes: Optional[int] = Field(None, description="循环间隔（分钟）")
    executed: bool = Field(default=False, description="是否已执行")
    world_id: str = Field(..., description="所属世界ID")


class TimePointModel(BaseModel):
    """时间点记录模型"""
    id: str = Field(..., description="时间点ID")
    world_id: str = Field(..., description="所属世界ID")
    time: datetime = Field(..., description="时间点")
    tick_count: int = Field(..., description="时钟周期数")
    note: str = Field(default="", description="备注说明")
    day_phase: DayPhaseEnum = Field(..., description="时段")
    season: SeasonEnum = Field(..., description="季节")
    created_at: datetime = Field(default_factory=datetime.utcnow, description="创建时间")


class TimeBranchModel(BaseModel):
    """时间分支模型"""
    id: str = Field(..., description="分支ID")
    world_id: str = Field(..., description="所属世界ID")
    name: str = Field(..., description="分支名称")
    time: datetime = Field(..., description="分支时间点")
    tick_count: int = Field(..., description="时钟周期数")
    created_at: datetime = Field(default_factory=datetime.utcnow, description="创建时间")
    is_active: bool = Field(default=False, description="是否为当前活跃分支")


class TimeUpdateModel(BaseModel):
    """时间更新模型"""
    world_id: str = Field(..., description="世界ID")
    current_time: datetime = Field(..., description="当前时间")
    previous_time: datetime = Field(..., description="之前时间")
    time_delta_minutes: float = Field(..., description="时间增量（分钟）")
    time_scale: float = Field(..., description="时间流速")
    tick_count: int = Field(..., description="时钟周期数")
    day_of_week: int = Field(..., description="星期几（0-6）")
    day_of_week_name: str = Field(..., description="星期名称")
    hour: int = Field(..., description="小时")
    minute: int = Field(..., description="分钟")
    day_phase: DayPhaseEnum = Field(..., description="时段")
    season: SeasonEnum = Field(..., description="季节")
    triggered_events: List[Dict[str, Any]] = Field(default_factory=list, description="触发的事件列表")


class TimeSystemConfig(BaseModel):
    """时间系统配置"""
    world_id: Optional[str] = Field(None, description="世界ID")
    time_scale: float = Field(default=1.0, ge=0.1, le=100.0, description="时间流速")
    time_mode: TimeModeEnum = Field(default=TimeModeEnum.LINEAR, description="时间模式")
    day_length: int = Field(default=24, ge=1, le=48, description="一天的小时数")
    enable_recording: bool = Field(default=True, description="是否启用时间记录")
    tick_interval_seconds: float = Field(default=1.0, ge=0.1, le=60.0, description="时钟周期间隔（秒）")


class TimeJumpRequest(BaseModel):
    """时间跳跃请求"""
    world_id: str = Field(..., description="世界ID")
    jump_type: str = Field(..., description="跳跃类型：forward/backward/to_absolute/to_year")
    target_time: Optional[datetime] = Field(None, description="目标时间（to_absolute）")
    delta_minutes: Optional[float] = Field(None, description="时间增量（分钟，forward/backward）")
    target_year: Optional[int] = Field(None, description="目标年份（to_year）")
    note: str = Field(default="", description="跳跃备注")


class TimeBranchRequest(BaseModel):
    """创建时间分支请求"""
    world_id: str = Field(..., description="世界ID")
    branch_name: str = Field(..., description="分支名称")
    note: str = Field(default="", description="分支说明")


class TimeControlRequest(BaseModel):
    """时间控制请求"""
    world_id: str = Field(..., description="世界ID")
    action: str = Field(..., description="操作：freeze/unfreeze/set_scale/advance")
    time_scale: Optional[float] = Field(None, description="新的时间流速（set_scale）")
    advance_minutes: Optional[float] = Field(None, description="推进分钟数（advance）")
    mode: Optional[TimeModeEnum] = Field(None, description="时间模式（设置模式）")


class TimeRecordRequest(BaseModel):
    """时间记录请求"""
    world_id: str = Field(..., description="世界ID")
    note: str = Field(..., description="记录备注")


class TimeHistoryResponse(BaseModel):
    """时间历史响应"""
    time_points: List[TimePointModel] = Field(default_factory=list, description="时间点列表")
    branches: List[TimeBranchModel] = Field(default_factory=list, description="时间分支列表")
    current_time: datetime = Field(..., description="当前时间")
    time_mode: TimeModeEnum = Field(..., description="当前时间模式")
    tick_count: int = Field(..., description="时钟周期数")


class WorldTimeLogModel(BaseModel):
    """世界时间日志模型（数据库存储）"""
    id: int = Field(default=None, description="主键ID")
    world_id: str = Field(..., description="世界ID")
    current_time: datetime = Field(..., description="当前时间")
    tick_number: int = Field(..., description="时钟周期编号")
    time_scale: float = Field(default=1.0, description="时间流速")
    day_phase: str = Field(..., description="时段")
    season: str = Field(..., description="季节")
    event_count: int = Field(default=0, description="该周期触发的事件数量")
    time_mode: str = Field(default="linear", description="时间模式")
    log_data: Dict[str, Any] = Field(default_factory=dict, description="额外的日志数据")
    created_at: datetime = Field(default_factory=datetime.utcnow, description="记录时间")
